File: kurrent_skills/templates/fsharp/Projection.fs
// KurrentDB F# In-Memory Projection Example
// Demonstrates: Subscribe to events, build state, track checkpoint

module Projection

open System
open System.Collections.Generic
open System.Text
open System.Text.Json
open System.Threading
open KurrentDB.Client

// === MINIMAL PROJECTION FRAMEWORK ===

type EventHandler<'TState> = 'TState -> JsonElement -> 'TState

type Projection<'TState when 'TState : (new : unit -> 'TState)> = {
    Name: string
    mutable State: Dictionary<string, 'TState>
    mutable Checkpoint: Position option
    Handlers: Dictionary<string, EventHandler<'TState>>
}

module Projection =
    let create<'TState when 'TState : (new : unit -> 'TState)> name : Projection<'TState> =
        {
            Name = name
            State = Dictionary<string, 'TState>()
            Checkpoint = None
            Handlers = Dictionary<string, EventHandler<'TState>>()
        }

    let on eventType (handler: EventHandler<'TState>) (projection: Projection<'TState>) =
        projection.Handlers.[eventType] <- handler
        projection

    let get streamId (projection: Projection<'TState>) =
        match projection.State.TryGetValue(streamId) with
        | true, state -> Some state
        | false, _ -> None

    let apply (resolved: ResolvedEvent) (projection: Projection<'TState>) =
        let evt = resolved.Event
        match projection.Handlers.TryGetValue(evt.EventType) with
        | true, handler ->
            let streamId = evt.EventStreamId
            let current =
                match projection.State.TryGetValue(streamId) with
                | true, s -> s
                | false, _ -> new 'TState()
            let data = JsonDocument.Parse(evt.Data.ToArray()).RootElement
            projection.State.[streamId] <- handler current data
            projection.Checkpoint <- Some evt.Position
            true
        | false, _ -> false

// === ORDER STATE ===

type OrderState() =
    member val OrderId = "" with get, set
    member val CustomerId = "" with get, set
    member val Amount = 0m with get, set
    member val Status = "unknown" with get, set
    member val Items = ResizeArray<string>() with get, set
    member val ShippedAt = "" with get, set

// === RUN PROJECTION ===

let run () =
    task {
        // === SETUP ===
        let connectionString =
            Environment.GetEnvironmentVariable("KURRENTDB_CONNECTION_STRING")
            |> Option.ofObj
            |> Option.defaultValue "kurrentdb://localhost:2113?tls=false"

        let settings = KurrentDBClientSettings.Create(connectionString)
        use client = new KurrentDBClient(settings)
        printfn "Connected to KurrentDB at %s" connectionString

        // Helper for safe property access
        let tryGetString (prop: string) (data: JsonElement) =
            match data.TryGetProperty(prop) with
            | true, el -> el.GetString()
            | false, _ -> ""
        let tryGetDecimal (prop: string) (data: JsonElement) =
            match data.TryGetProperty(prop) with
            | true, el -> el.GetDecimal()
            | false, _ -> 0m

        // === DEFINE PROJECTION ===
        let orderProjection =
            Projection.create<OrderState> "OrderSummary"
            |> Projection.on "OrderCreated" (fun _ data ->
                let state = OrderState()
                state.OrderId <- tryGetString "orderId" data
                state.CustomerId <- tryGetString "customerId" data
                state.Amount <- tryGetDecimal "amount" data
                state.Status <- "created"
                state)
            |> Projection.on "ItemAdded" (fun state data ->
                match data.TryGetProperty("item") with
                | true, el -> state.Items.Add(el.GetString())
                | false, _ -> ()
                state.Amount <- state.Amount + tryGetDecimal "price" data
                state)
            |> Projection.on "OrderShipped" (fun state data ->
                state.Status <- "shipped"
                state.ShippedAt <- tryGetString "shippedAt" data
                state)
            |> Projection.on "OrderCompleted" (fun state _ ->
                state.Status <- "completed"
                state)

        // === TEST: Append test events ===
        printfn "\n=== Appending test events ==="

        let orderId1 = Guid.NewGuid().ToString()
        let orderId2 = Guid.NewGuid().ToString()
        let stream1 = sprintf "order-%s" orderId1
        let stream2 = sprintf "order-%s" orderId2

        let makeEvent eventType (data: obj) =
            EventData(
                Uuid.NewUuid(),
                eventType,
                JsonSerializer.SerializeToUtf8Bytes(data)
            )

        // Order 1: Created -> ItemAdded -> Shipped -> Completed
        let! _ = client.AppendToStreamAsync(stream1, StreamState.Any, [|
            makeEvent "OrderCreated" {| orderId = orderId1; customerId = "cust-1"; amount = 100m |}
        |])
        let! _ = client.AppendToStreamAsync(stream1, StreamState.Any, [|
            makeEvent "ItemAdded" {| item = "Widget"; price = 25m |}
        |])
        let! _ = client.AppendToStreamAsync(stream1, StreamState.Any, [|
            makeEvent "OrderShipped" {| shippedAt = "2024-01-15T10:00:00Z" |}
        |])
        let! _ = client.AppendToStreamAsync(stream1, StreamState.Any, [|
            makeEvent "OrderCompleted" {||}
        |])

        // Order 2: Created -> ItemAdded (still pending)
        let! _ = client.AppendToStreamAsync(stream2, StreamState.Any, [|
            makeEvent "OrderCreated" {| orderId = orderId2; customerId = "cust-2"; amount = 50m |}
        |])
        let! _ = client.AppendToStreamAsync(stream2, StreamState.Any, [|
            makeEvent "ItemAdded" {| item = "Gadget"; price = 30m |}
        |])

        printfn "Created order streams: %s, %s" stream1 stream2

        // === RUN PROJECTION ===
        printfn "\n=== Running projection ==="

        let filterOptions = SubscriptionFilterOptions(EventTypeFilter.ExcludeSystemEvents())
        use cts = new CancellationTokenSource(TimeSpan.FromSeconds(30.0))

        use subscription = client.SubscribeToAll(FromAll.Start, filterOptions = filterOptions)
        let enumerator = subscription.Messages.GetAsyncEnumerator(cts.Token)

        let mutable processedCount = 0
        let targetStreams = HashSet([stream1; stream2])
        let targetEventsCount = Dictionary<string, int>()
        targetEventsCount.[stream1] <- 0
        targetEventsCount.[stream2] <- 0
        let expectedCounts = dict [stream1, 4; stream2, 2] // Order 1: 4 events, Order 2: 2 events
        let mutable continueLoop = true

        while continueLoop do
            let! moveNext = enumerator.MoveNextAsync()
            if moveNext then
                match enumerator.Current with
                | :? StreamMessage.Event as evtMsg ->
                    let resolved = evtMsg.ResolvedEvent
                    if Projection.apply resolved orderProjection then
                        processedCount <- processedCount + 1
                        let streamId = resolved.Event.EventStreamId
                        printfn "  Processed: %s on %s" resolved.Event.EventType streamId

                        if targetStreams.Contains(streamId) then
                            targetEventsCount.[streamId] <- targetEventsCount.[streamId] + 1

                    // Stop when we've processed all test events for both streams
                    if targetEventsCount.[stream1] >= expectedCounts.[stream1] &&
                       targetEventsCount.[stream2] >= expectedCounts.[stream2] then
                        continueLoop <- false

                    // Safety limit
                    if processedCount > 200 then
                        continueLoop <- false
                | _ -> ()
            else
                continueLoop <- false

        // === VERIFY RESULTS ===
        printfn "\n=== Projection Results ==="

        let order1State = Projection.get stream1 orderProjection
        let order2State = Projection.get stream2 orderProjection

        match order1State with
        | Some s ->
            printfn "\nOrder 1 (%s):" stream1
            printfn "  OrderId: %s" s.OrderId
            printfn "  CustomerId: %s" s.CustomerId
            printfn "  Amount: %M" s.Amount
            printfn "  Status: %s" s.Status
            printfn "  Items: %A" (s.Items |> Seq.toList)
        | None -> printfn "\nOrder 1 not found"

        match order2State with
        | Some s ->
            printfn "\nOrder 2 (%s):" stream2
            printfn "  OrderId: %s" s.OrderId
            printfn "  CustomerId: %s" s.CustomerId
            printfn "  Amount: %M" s.Amount
            printfn "  Status: %s" s.Status
            printfn "  Items: %A" (s.Items |> Seq.toList)
        | None -> printfn "\nOrder 2 not found"

        // === ASSERTIONS ===
        printfn "\n=== Running assertions ==="

        let mutable passed = true

        // Order 1 assertions
        match order1State with
        | Some s ->
            if s.Status <> "completed" then
                printfn "FAIL: Order 1 status should be 'completed', got '%s'" s.Status
                passed <- false
            if s.Amount <> 125m then
                printfn "FAIL: Order 1 amount should be 125, got %M" s.Amount
                passed <- false
            if s.Items.Count <> 1 then
                printfn "FAIL: Order 1 should have 1 item, got %d" s.Items.Count
                passed <- false
        | None ->
            printfn "FAIL: Order 1 state not found"
            passed <- false

        // Order 2 assertions
        match order2State with
        | Some s ->
            if s.Status <> "created" then
                printfn "FAIL: Order 2 status should be 'created', got '%s'" s.Status
                passed <- false
            if s.Amount <> 80m then
                printfn "FAIL: Order 2 amount should be 80, got %M" s.Amount
                passed <- false
        | None ->
            printfn "FAIL: Order 2 state not found"
            passed <- false

        // Checkpoint assertion
        if orderProjection.Checkpoint.IsNone then
            printfn "FAIL: Checkpoint should be set"
            passed <- false

        if passed then
            printfn "\nAll projection tests passed!"
            return 0
        else
            printfn "\nSome tests failed!"
            return 1
    }

File: kurrent_skills/templates/dotnet/Projection.cs
// KurrentDB .NET In-Memory Projection Example
// Demonstrates: Subscribe to events, build state, track checkpoint

using System.Text;
using System.Text.Json;
using KurrentDB.Client;

// === MINIMAL PROJECTION FRAMEWORK ===
public class Projection<TState> where TState : class, new()
{
    public string Name { get; }
    public Dictionary<string, TState> State { get; } = new();
    public Position? Checkpoint { get; private set; }

    private readonly Dictionary<string, Func<TState, JsonElement, TState>> _handlers = new();

    public Projection(string name) => Name = name;

    public Projection<TState> On(string eventType, Func<TState, JsonElement, TState> handler)
    {
        _handlers[eventType] = handler;
        return this;
    }

    public TState? Get(string streamId) => State.GetValueOrDefault(streamId);

    public Dictionary<string, TState> GetAll() => new(State);

    public bool Apply(ResolvedEvent resolved)
    {
        var evt = resolved.Event;
        if (!_handlers.TryGetValue(evt.EventType, out var handler))
            return false;

        var streamId = evt.EventStreamId;
        var current = State.GetValueOrDefault(streamId) ?? new TState();
        var data = JsonDocument.Parse(evt.Data.ToArray()).RootElement;
        State[streamId] = handler(current, data);
        Checkpoint = evt.Position;
        return true;
    }
}

// === ORDER STATE ===
public class OrderState
{
    public string? OrderId { get; set; }
    public string? CustomerId { get; set; }
    public decimal Amount { get; set; }
    public string Status { get; set; } = "unknown";
    public List<string> Items { get; set; } = new();
    public string? ShippedAt { get; set; }
}

// === MAIN ===
public static class ProjectionExample
{
    public static async Task Run()
    {
        // === SETUP ===
        var connectionString = Environment.GetEnvironmentVariable("KURRENTDB_CONNECTION_STRING")
            ?? "kurrentdb://localhost:2113?tls=false";

        var settings = KurrentDBClientSettings.Create(connectionString);
        await using var client = new KurrentDBClient(settings);
        Console.WriteLine($"Connected to KurrentDB at {connectionString}");

        // === DEFINE PROJECTION ===
        var orderProjection = new Projection<OrderState>("OrderSummary")
            .On("OrderCreated", (state, data) => new OrderState
            {
                OrderId = data.TryGetProperty("orderId", out var oid) ? oid.GetString() : null,
                CustomerId = data.TryGetProperty("customerId", out var cid) ? cid.GetString() : null,
                Amount = data.TryGetProperty("amount", out var amt) ? amt.GetDecimal() : 0m,
                Status = "created",
                Items = new List<string>()
            })
            .On("ItemAdded", (state, data) =>
            {
                if (data.TryGetProperty("item", out var item))
                    state.Items.Add(item.GetString()!);
                if (data.TryGetProperty("price", out var price))
                    state.Amount += price.GetDecimal();
                return state;
            })
            .On("OrderShipped", (state, data) =>
            {
                state.Status = "shipped";
                state.ShippedAt = data.TryGetProperty("shippedAt", out var sa) ? sa.GetString() : null;
                return state;
            })
            .On("OrderCompleted", (state, _) =>
            {
                state.Status = "completed";
                return state;
            });

        // === TEST: Append test events ===
        Console.WriteLine("\n=== Appending test events ===");

        var orderId1 = Guid.NewGuid().ToString();
        var orderId2 = Guid.NewGuid().ToString();
        var stream1 = $"order-{orderId1}";
        var stream2 = $"order-{orderId2}";

        EventData MakeEvent(string type, object data) => new(
            Uuid.NewUuid(),
            type,
            Encoding.UTF8.GetBytes(JsonSerializer.Serialize(data))
        );

        // Order 1: Created -> ItemAdded -> Shipped -> Completed
        await client.AppendToStreamAsync(stream1, StreamState.Any, new[] {
            MakeEvent("OrderCreated", new { orderId = orderId1, customerId = "cust-1", amount = 100m })
        });
        await client.AppendToStreamAsync(stream1, StreamState.Any, new[] {
            MakeEvent("ItemAdded", new { item = "Widget", price = 25m })
        });
        await client.AppendToStreamAsync(stream1, StreamState.Any, new[] {
            MakeEvent("OrderShipped", new { shippedAt = "2024-01-15T10:00:00Z" })
        });
        await client.AppendToStreamAsync(stream1, StreamState.Any, new[] {
            MakeEvent("OrderCompleted", new { })
        });

        // Order 2: Created -> ItemAdded (still pending)
        await client.AppendToStreamAsync(stream2, StreamState.Any, new[] {
            MakeEvent("OrderCreated", new { orderId = orderId2, customerId = "cust-2", amount = 50m })
        });
        await client.AppendToStreamAsync(stream2, StreamState.Any, new[] {
            MakeEvent("ItemAdded", new { item = "Gadget", price = 30m })
        });

        Console.WriteLine($"Created order streams: {stream1}, {stream2}");

        // === RUN PROJECTION ===
        Console.WriteLine("\n=== Running projection ===");

        var processedCount = 0;
        var targetStreams = new HashSet<string> { stream1, stream2 };
        var targetEventsCount = new Dictionary<string, int> { { stream1, 0 }, { stream2, 0 } };
        var expectedCounts = new Dictionary<string, int> { { stream1, 4 }, { stream2, 2 } };

        using var cts = new CancellationTokenSource(TimeSpan.FromSeconds(30));
        var filterOptions = new SubscriptionFilterOptions(EventTypeFilter.ExcludeSystemEvents());

        await using var subscription = client.SubscribeToAll(FromAll.Start, filterOptions: filterOptions);

        await foreach (var message in subscription.Messages.WithCancellation(cts.Token))
        {
            if (message is StreamMessage.Event evtMsg)
            {
                var resolved = evtMsg.ResolvedEvent;
                if (orderProjection.Apply(resolved))
                {
                    processedCount++;
                    Console.WriteLine($"  Processed: {resolved.Event.EventType} on {resolved.Event.EventStreamId}");

                    if (targetStreams.Contains(resolved.Event.EventStreamId))
                        targetEventsCount[resolved.Event.EventStreamId]++;
                }

                // Stop when we've processed our test events
                if (targetEventsCount[stream1] >= expectedCounts[stream1] &&
                    targetEventsCount[stream2] >= expectedCounts[stream2])
                    break;

                // Safety limit
                if (processedCount > 100) break;
            }
        }

        // === VERIFY RESULTS ===
        Console.WriteLine("\n=== Projection Results ===");

        var order1State = orderProjection.Get(stream1);
        var order2State = orderProjection.Get(stream2);

        Console.WriteLine($"\nOrder 1 ({stream1}):");
        Console.WriteLine(JsonSerializer.Serialize(order1State, new JsonSerializerOptions { WriteIndented = true }));

        Console.WriteLine($"\nOrder 2 ({stream2}):");
        Console.WriteLine(JsonSerializer.Serialize(order2State, new JsonSerializerOptions { WriteIndented = true }));

        // === ASSERTIONS ===
        Console.WriteLine("\n=== Running assertions ===");

        var passed = true;

        // Order 1 assertions
        if (order1State?.Status != "completed")
        {
            Console.WriteLine($"FAIL: Order 1 status should be 'completed', got '{order1State?.Status}'");
            passed = false;
        }
        if (order1State?.Amount != 125m)
        {
            Console.WriteLine($"FAIL: Order 1 amount should be 125, got {order1State?.Amount}");
            passed = false;
        }
        if (order1State?.Items.Count != 1)
        {
            Console.WriteLine($"FAIL: Order 1 should have 1 item, got {order1State?.Items.Count}");
            passed = false;
        }

        // Order 2 assertions
        if (order2State?.Status != "created")
        {
            Console.WriteLine($"FAIL: Order 2 status should be 'created', got '{order2State?.Status}'");
            passed = false;
        }
        if (order2State?.Amount != 80m)
        {
            Console.WriteLine($"FAIL: Order 2 amount should be 80, got {order2State?.Amount}");
            passed = false;
        }

        // Checkpoint assertion
        if (orderProjection.Checkpoint == null)
        {
            Console.WriteLine("FAIL: Checkpoint should be set");
            passed = false;
        }

        if (passed)
        {
            Console.WriteLine("\nAll projection tests passed!");
        }
        else
        {
            Console.WriteLine("\nSome tests failed!");
            Environment.Exit(1);
        }
    }
}
